File: src/main/java/io/github/intellij/dlanguage/psi/references/DReferenceProvider.java
package io.github.intellij.dlanguage.psi.references;

import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiNamedElement;
import com.intellij.psi.PsiReference;
import com.intellij.psi.PsiReferenceProvider;
import com.intellij.util.ProcessingContext;
import io.github.intellij.dlanguage.DLanguage;
import org.jetbrains.annotations.NotNull;

public class DReferenceProvider extends PsiReferenceProvider {
    @NotNull
    @Override
    public PsiReference[] getReferencesByElement(@NotNull final PsiElement element,
                                                 @NotNull final ProcessingContext context) {
        if (!element.getLanguage().is(DLanguage.INSTANCE)) {
            return PsiReference.EMPTY_ARRAY;
        }

        if (element instanceof PsiNamedElement) {
            final PsiNamedElement se = (PsiNamedElement) element;
            return new PsiReference[]{new DReference(se, se.getTextRange())};
        }
        return PsiReference.EMPTY_ARRAY;
    }
}

File: src/main/kotlin/io/github/intellij/dlanguage/psi/references/DReferenceProvider.kt
package io.github.intellij.dlanguage.psi.references

import com.intellij.psi.PsiElement
import com.intellij.psi.PsiNamedElement
import com.intellij.psi.PsiReference
import com.intellij.psi.PsiReferenceProvider
import com.intellij.util.ProcessingContext
import io.github.intellij.dlanguage.DLanguage

class DReferenceProvider : PsiReferenceProvider() {
    override fun getReferencesByElement(
        element: PsiElement,
        context: ProcessingContext
    ): Array<PsiReference> {
        if (!element.language.`is`(DLanguage)) {
            return PsiReference.EMPTY_ARRAY
        }

        if (element is PsiNamedElement) {
            return arrayOf(DReference(element, element.textRange))
        }
        return PsiReference.EMPTY_ARRAY
    }
}
